src/storage/redis_storage.rs

```diff
@@ -1,34 +1,247 @@
+        let value: String = serde_json::to_string(&msg)?;
+
+        // add to backlog
+        let key: String = format!("{}.backlog.{}", &self.prefix, &msg.uid);
+        conn.set_ex(&key, &value, self.ttl_in_seconds).await?;
+
+        // push to forward for every destination
+        let queue: String = format!("{}.system.forward", &self.prefix);
+        for dst in msg.dst {
+            let value: String = format!("{}.{}", &msg.uid, &dst.to_string());
+            conn.rpush(&queue, &value).await?
+        }
+
+        Ok(())
     }
 
-    async fn reply(&self, _msg: Message) -> anyhow::Result<()> {
-        unimplemented!()
+    async fn reply(&self, msg: Message) -> Result<()> {
+        let mut conn = self.get_connection().await?;
+        let value: String = serde_json::to_string(&msg)?;
+
+        conn.rpush(&msg.ret, &value).await?;
+        Ok(())
     }
 
-    async fn local(&self) -> anyhow::Result<Message> {
-        unimplemented!()
+    async fn local(&self) -> Result<Message> {
+        let mut conn = self.get_connection().await?;
+        let queue: String = format!("{}.system.local", &self.prefix);
+        let ret: Option<Vec<String>> = conn.blpop(&queue, 0).await?;
+
+        let vec = ret.unwrap();
+        if let [_, value] = &vec[..] {
+            let msg: Message = serde_json::from_str(&value)?;
+            Ok(msg)
+        } else {
+            // better to return an error? of what type?
+            unreachable!()
+        }
     }
 
-    async fn queued(&self) -> anyhow::Result<crate::types::QueuedMessage> {
-        Ok(QueuedMessage::Forward(Message::default()))
+    async fn queued(&self) -> Result<Option<QueuedMessage>> {
+        let mut conn = self.get_connection().await?;
+        let forward_queue: String = format!("{}.system.forward", &self.prefix);
+        let reply_queue: String = format!("{}.system.reply", &self.prefix);
+
+        let queues: Vec<String> = vec![forward_queue, reply_queue];
+        let ret: Option<Vec<String>> = conn.blpop(&queues, 0).await?;
+
+        let vec = ret.unwrap();
+        if let [queue, value] = &vec[..] {
+            match queue {
+                forward_queue => {
+                    // forward queue has only <id>.<dst> pair
+                    // a better way to parse these pair?
+                    let mut iter = value.splitn(2, '.');
+                    let id: String = String::from(iter.next().unwrap());
+                    let dst: usize = iter.next().unwrap().parse().unwrap();
+                    //  message can expire
+                    let ret: Option<Message> = self.get(id).await?;
+                    match ret {
+                        Some(mut msg) => {
+                            msg.dst = vec![dst];
+                            Ok(Some(QueuedMessage::Forward(msg)))
+                        }
+                        None => {
+                            // what to do in case the message is gone!
+                            Ok(None)
```

Comment:
As per the specs, this method need to try in a loop, if a message is not available anymore. you try to pop the next, and so on until you hit a valid message. It means that this method need to always return `Result<QueuedMessage>` not `Result<Option<QueuedMessage>>`